lib/net.js

```diff
@@ -687,6 +666,49 @@ Socket.prototype._write = function(data, encoding, cb) {
   this._writeGeneric(false, data, encoding, cb);
 };
 
+function _end(socket, cb) {
+  debug('_end');
+
+  // If still connecting - defer handling 'finish' until 'connect' will happen
+  if (socket._connecting) {
+    debug('_end: not yet connected');
+    return socket.once('connect', function() {
+      socket._end(cb);
+    });
+  }
+
+  if (!socket.readable || socket._readableState.ended) {
+    debug('_end: not readable or ended');
+    return cb();
+  }
+
+  // otherwise, just shutdown, or destroy() if not possible
+  if (!socket._handle || !socket._handle.shutdown) {
+    debug('_end: no handle or handle does not support shutdown');
+    return cb();
+  }
+
+  var req = new ShutdownWrap();
+  req.oncomplete = afterShutdown;
+  req.flushCb = cb;
+  var err = socket._handle.shutdown(req);
+
+  if (err) {
+    debug('_end: errno %s', err);
+    return cb(errnoException(err, 'shutdown'));
+  }
+}
+
+
+function afterShutdown(status, handle, req) {
+  var self = handle.owner;
+
+  debug('afterShutdown destroyed=%j', self.destroyed,
+        self._readableState);
+
+  req.flushCb();
```

Comment:
Inconsistency: `cb` runs with the global object as the receiver everywhere else but here `req` is the receiver.